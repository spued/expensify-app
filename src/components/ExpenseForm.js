import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';


export default class ExpenseForm extends React.Component {
    constructor(props){
        super(props);
        this.state = {
            description : props.expense ? props.expense.description : '',
            note : props.expense ? props.expense.note : '',
            amount : props.expense ? (props.expense.amount/100).toString() : '',
            createdAt : props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused : false,
            error_message : ''
        };
    }
    
    onDescriptionChange = (evt) => {
        const description = evt.target.value;
        this.setState (() => ({ description }));
    };
    onNoteChange = (evt) => {
        const note = evt.target.value;
        this.setState (() => ({ note }));
    };
    onAmountChange = (evt) => {
        const amount = evt.target.value;
        if(amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState (() => ({ amount }));
        }
    };
    onDateChange = (createdAt) => {
        if(createdAt) {
            this.setState (() => ({ createdAt }));
        }
    };
    onFocusChanged = ({ focused }) => {
        this.setState (() => ({ calendarFocused : focused }));
    };
    onSubmit = (evt) => {
        evt.preventDefault();
        if(!this.state.description || !this.state.amount) {
            this.setState (() => ({ error_message : 'I need description and amount.' }));
        } else {
            this.setState (() => ({ error_message : '' }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount,10),
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            })
        }
    }
    onRemove = () => {
        this.props.onRemove();
    }
    render() {
        return (
            <div>
                {this.state.error_message && <p>{this.state.error_message}</p>}
                <form onSubmit={this.onSubmit}> 
                    <input 
                        type='text'
                        autoFocus
                        placeholder='Description'
                        value={this.state.description}
                        onChange={this.onDescriptionChange}>
                    </input>
                    <input 
                        type='number'
                        placeholder='Amount'
                        value={this.state.amount}
                        onChange={this.onAmountChange}>
                    </input>
                    <SingleDatePicker
                        date = {this.state.createdAt}
                        onDateChange = {this.onDateChange}
                        focused = { this.state.calendarFocused }
                        onFocusChange = { this.onFocusChanged }
                        numberOfMonths = {1}
                        isOutsideRange= {() => false }
                    />
                    <textarea
                        placeholder='Add the note for your expense. (Optional)'
                        value={this.state.note}
                        onChange={this.onNoteChange}>
                    </textarea>
                    <button>Add Expense</button>
                </form>
                <button id='remove_button' onClick={this.onRemove}>Remove</button>
            </div>
        )
    };      
}
