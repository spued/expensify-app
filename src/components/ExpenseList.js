import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses'

export const ExpenseList = (props) => (
    <div>
       {
            props.expenses.length === 0 ? (
                <p> No expense.</p>
            ) : (
                props.expenses.map((expense) => {
                    return <ExpenseListItem key={ expense.id } { ...expense }/>;
                }) 
            )
        }
    </div>
);

const mapStateToProps = (state) => {
    let _expenses = selectExpenses(state.expenses, state.filters);
    return {
        expenses: _expenses
    }
}

export default connect(mapStateToProps)(ExpenseList);