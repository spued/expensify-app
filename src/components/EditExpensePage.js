import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';

// make it class
export class EditExpensePage extends React.Component { 
    onSubmit = (expense) => {
        //console.log(expense);
        this.props.editExpense( this.props.expense.id, expense);
        this.props.history.push('/');
    }
    onRemove = () => {
        this.props.removeExpense({ id : this.props.expense.id});
        this.props.history.push('/');
    }
    render() {
        return (
            <div>
                This is edit expense component id = { this.props.expense.id}
                <ExpenseForm 
                    expense = { this.props.expense }
                    onSubmit = { this.onSubmit }
                />
                <button onClick={this.onRemove}></button>
            </div>
        )
    }
}
const mapDispatchToProps = (dispatch, props) => ({
    editExpense: (id, expense) => dispatch(editExpense(id, expense)),
    removeExpense : (id) => dispatch(removeExpense({ id }))
})
const mapStateToProps = (state, props) => ({
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
})
export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);