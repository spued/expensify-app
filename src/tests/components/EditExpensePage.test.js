import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from '../../components/EditExpensePage';
import expense from '../fixtures/expenses';

let startRemoveExpense, startEditExpense, history, wrapper;

beforeEach(() => {
    startEditExpense = jest.fn();
    startRemoveExpense = jest.fn();
    history = { push: jest.fn( )};
    wrapper = shallow(<EditExpensePage 
        expense={ expense[2] } 
        startEditExpense={startEditExpense} 
        startRemoveExpense={startRemoveExpense} 
        history={history} 
        />);
})

test('Should render edit expense page correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('Should edit expense correctly', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expense[2]);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(startEditExpense).toHaveBeenLastCalledWith(expense[2].id, expense[2]);
});

test('Should remove expense page correctly', () => {
    wrapper.find('button').simulate('click');
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(startRemoveExpense).toHaveBeenLastCalledWith({ 
        id : expense[2].id 
    });
});