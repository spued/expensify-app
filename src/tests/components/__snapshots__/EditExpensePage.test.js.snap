// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render edit expense page correctly 1`] = `
<div>
  This is edit expense component id = 
  3
  <ExpenseForm
    expense={
      {
        "amount": 4500,
        "createdAt": 345600000,
        "description": "Credit cards",
        "id": "3",
        "note": "",
      }
    }
    onRemove={[Function]}
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  />
</div>
`;
